File: back-end/src/main/java/es/ull/project/domain/valueobject/policy/ServicePolicies.java
package es.ull.project.domain.valueobject.policy;

import java.util.Objects;
import java.util.Optional;

/**
 * ServicePolicies
 *
 * Represents a group of service-related policies and constraints.
 * This value object aggregates coherent policy parameters.
 */
public final class ServicePolicies {

    private static final String ERROR_NEGATIVE_VALUE =
            "Policy values cannot be negative";
    private static final int ZERO = 0;

    /**
     * Optional.
     * Maximum service distance allowed. Measured in meters.
     * @optional
     */
    private final Double maxServiceDistance;

    /**
     * Optional.
     * Maximum service time allowed. Measured in minutes.
     * @optional
     */
    private final Integer maxServiceTime;

    /**
     * Optional.
     * Maximum number of infrastructures allowed.
     * @optional
     */
    private final Integer maxInfrastructureCount;

    /**
     * Optional.
     * Maximum emission limit. Measured in kilograms of CO2.
     * @optional
     */
    private final Double maxEmissions;

    /**
     * Creates a new ServicePolicies value object.
     *
     * @param maxServiceDistance     Maximum service distance (meters).
     * @param maxServiceTime         Maximum service time (minutes).
     * @param maxInfrastructureCount Maximum number of infrastructures.
     * @param maxEmissions           Maximum emissions allowed (kg CO2).
     */
    public ServicePolicies(
            Double maxServiceDistance,
            Integer maxServiceTime,
            Integer maxInfrastructureCount,
            Double maxEmissions
    ) {
        validateNonNegative(maxServiceDistance);
        validateNonNegative(maxServiceTime);
        validateNonNegative(maxInfrastructureCount);
        validateNonNegative(maxEmissions);
        this.maxServiceDistance = maxServiceDistance;
        this.maxServiceTime = maxServiceTime;
        this.maxInfrastructureCount = maxInfrastructureCount;
        this.maxEmissions = maxEmissions;
    }

    /**
     * Validates that a numeric policy value is non-negative.
     *
     * @param value Value to validate.
     */
    private void validateNonNegative(Number value) {
        if (value != null && value.doubleValue() < ZERO) {
            throw new IllegalArgumentException(ERROR_NEGATIVE_VALUE);
        }
    }

    /**
     * Returns the maximum service distance, if defined.
     *
     * @return Optional maximum service distance.
     */
    public Optional<Double> getMaxServiceDistance() {
        return Optional.ofNullable(this.maxServiceDistance);
    }

    /**
     * Returns the maximum service time, if defined.
     *
     * @return Optional maximum service time.
     */
    public Optional<Integer> getMaxServiceTime() {
        return Optional.ofNullable(this.maxServiceTime);
    }

    /**
     * Returns the maximum number of infrastructures, if defined.
     *
     * @return Optional maximum infrastructure count.
     */
    public Optional<Integer> getMaxInfrastructureCount() {
        return Optional.ofNullable(this.maxInfrastructureCount);
    }

    /**
     * Returns the maximum emissions allowed, if defined.
     *
     * @return Optional emissions limit.
     */
    public Optional<Double> getMaxEmissions() {
        return Optional.ofNullable(this.maxEmissions);
    }

    /**
     * Checks whether any policy is defined.
     *
     * @return True if at least one policy is present.
     */
    public boolean hasAnyPolicy() {
        return maxServiceDistance != null
                || maxServiceTime != null
                || maxInfrastructureCount != null
                || maxEmissions != null;
    }

    /**
     * Validates whether a service assignment complies with defined policies.
     *
     * @param serviceDistance Distance of the service assignment.
     * @param serviceTime Service time of the assignment.
     * @return Optional with error message if validation fails, empty if compliant.
     */
    public Optional<String> validateServiceAssignment(double serviceDistance, int serviceTime) {
        if (maxServiceDistance != null && serviceDistance > maxServiceDistance) {
            return Optional.of(String.format(
                    "Service distance %.2f km exceeds maximum allowed %.2f km",
                    serviceDistance, maxServiceDistance
            ));
        }
        if (maxServiceTime != null && serviceTime > maxServiceTime) {
            return Optional.of(String.format(
                    "Service time %d minutes exceeds maximum allowed %d minutes",
                    serviceTime, maxServiceTime
            ));
        }
        return Optional.empty();
    }

    /**
     * Checks if all policy constraints are satisfied.
     *
     * @param serviceDistance Distance of the service.
     * @param serviceTime Service time.
     * @param infrastructureCount Number of infrastructures.
     * @param emissions Emissions value.
     * @return True if compliant with all policies, false otherwise.
     */
    public boolean isCompliant(
            double serviceDistance,
            int serviceTime,
            int infrastructureCount,
            double emissions
    ) {
        if (maxServiceDistance != null && serviceDistance > maxServiceDistance) {
            return false;
        }
        if (maxServiceTime != null && serviceTime > maxServiceTime) {
            return false;
        }
        if (maxInfrastructureCount != null && infrastructureCount > maxInfrastructureCount) {
            return false;
        }
        if (maxEmissions != null && emissions > maxEmissions) {
            return false;
        }
        return true;
    }

    /**
     * Compares this service policies with another object for equality.
     *
     * @param otherObject the object to compare with
     * @return true if the objects are equal, false otherwise
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        ServicePolicies other = (ServicePolicies) otherObject;
        return Objects.equals(this.maxServiceDistance, other.maxServiceDistance)
                && Objects.equals(this.maxServiceTime, other.maxServiceTime)
                && Objects.equals(this.maxInfrastructureCount, other.maxInfrastructureCount)
                && Objects.equals(this.maxEmissions, other.maxEmissions);
    }

    /**
     * Returns the hash code of this service policies.
     *
     * @return the hash code
     */
    @Override
    public int hashCode() {
        return Objects.hash(
                maxServiceDistance,
                maxServiceTime,
                maxInfrastructureCount,
                maxEmissions
        );
    }

    /**
     * Returns a string representation of this service policies.
     *
     * @return a formatted string with policy details
     */
    @Override
    public String toString() {
        return String.format(
                "ServicePolicies={maxServiceDistance=%s, maxServiceTime=%s, maxInfrastructureCount=%s, maxEmissions=%s}",
                maxServiceDistance,
                maxServiceTime,
                maxInfrastructureCount,
                maxEmissions
        );
    }
}

File: front-end/src/domain/valueobject/policy/service-policies.ts
/**
 * ServicePolicies
 *
 * Aggregates optional policy parameters for services.
 */
export class ServicePolicies {
  /**
   * Error message for negative policy values.
   */
  private static readonly ERROR_NEGATIVE = 'Policy values cannot be negative';

  /**
   * Maximum service distance in meters (nullable).
   */
  readonly maxServiceDistance?: number | null;

  /**
   * Maximum service time in minutes (nullable).
   */
  readonly maxServiceTime?: number | null;

  /**
   * Maximum infrastructure count (nullable).
   */
  readonly maxInfrastructureCount?: number | null;

  /**
   * Maximum emissions allowed (nullable).
   */
  readonly maxEmissions?: number | null;

  /**
   * Create policy container with optional numeric limits.
   * @param maxServiceDistance maximum distance in meters (nullable)
   * @param maxServiceTime maximum time in minutes (nullable)
   * @param maxInfrastructureCount maximum infrastructure count (nullable)
   * @param maxEmissions maximum emissions allowed (nullable)
   */
  constructor(maxServiceDistance?: number | null, maxServiceTime?: number | null, maxInfrastructureCount?: number | null, maxEmissions?: number | null) {
    this.validateNonNegative(maxServiceDistance);
    this.validateNonNegative(maxServiceTime);
    this.validateNonNegative(maxInfrastructureCount);
    this.validateNonNegative(maxEmissions);
    this.maxServiceDistance = maxServiceDistance ?? null;
    this.maxServiceTime = maxServiceTime ?? null;
    this.maxInfrastructureCount = maxInfrastructureCount ?? null;
    this.maxEmissions = maxEmissions ?? null;
  }

  /**
   * Validate that a provided numeric policy is non-negative.
   * @param value the value to validate
   * @throws Error if value is negative
   */
  private validateNonNegative(value?: number | null) {
    if (value != null && value < 0) throw new Error(ServicePolicies.ERROR_NEGATIVE);
  }

  /**
   * Return true when any policy value is present.
   * @returns true if any policy is set, false otherwise
   */
  hasAnyPolicy(): boolean {
    return this.maxServiceDistance != null || this.maxServiceTime != null || this.maxInfrastructureCount != null || this.maxEmissions != null;
  }

  /**
   * Validate a proposed service assignment against distance/time policies.
   * @param serviceDistance distance to check
   * @param serviceTime time to check
   * @returns null when valid, otherwise a human-readable violation message
   */
  validateServiceAssignment(serviceDistance: number, serviceTime: number): string | null {
    if (this.maxServiceDistance != null && serviceDistance > this.maxServiceDistance) {
      return `Service distance ${serviceDistance} exceeds maximum allowed ${this.maxServiceDistance}`;
    }
    if (this.maxServiceTime != null && serviceTime > this.maxServiceTime) {
      return `Service time ${serviceTime} minutes exceeds maximum allowed ${this.maxServiceTime} minutes`;
    }
    return null;
  }

  /**
   * Check compliance against all available policies.
   * @param serviceDistance the service distance to check
   * @param serviceTime the service time to check
   * @param infrastructureCount the infrastructure count to check
   * @param emissions the emissions to check
   * @returns true when all applicable policies are satisfied, false otherwise
   */
  isCompliant(serviceDistance: number, serviceTime: number, infrastructureCount: number, emissions: number): boolean {
    if (this.maxServiceDistance != null && serviceDistance > this.maxServiceDistance) return false;
    if (this.maxServiceTime != null && serviceTime > this.maxServiceTime) return false;
    if (this.maxInfrastructureCount != null && infrastructureCount > this.maxInfrastructureCount) return false;
    if (this.maxEmissions != null && emissions > this.maxEmissions) return false;
    return true;
  }

  /**
   * Equality by value of all fields.
   * @param other the object to compare with
   * @returns true if all fields are equal, false otherwise
   */
  equals(other: unknown): boolean {
    if (this === other) return true;
    if (!(other instanceof ServicePolicies)) return false;
    return this.maxServiceDistance === other.maxServiceDistance && this.maxServiceTime === other.maxServiceTime && this.maxInfrastructureCount === other.maxInfrastructureCount && this.maxEmissions === other.maxEmissions;
  }

  /**
   * Human-readable representation.
   * @returns a string representation of the ServicePolicies object
   */
  toString(): string {
    return `ServicePolicies={maxServiceDistance=${this.maxServiceDistance}, maxServiceTime=${this.maxServiceTime}, maxInfrastructureCount=${this.maxInfrastructureCount}, maxEmissions=${this.maxEmissions}}`;
  }
}
